### PR TITLE
gh-124111: Update tkinter for compatibility with Tcl/Tk 9.0.0

### PR DESCRIPTION
**Pull request by culler**
This PR updates tkinter, ttk, test.test_tkinter and test.test_ttk to make them work and pass tests when built against either TclTk 8.6.15, which was released today, or TclTk 9.0.0 which currently has a release candidate and will be released soon.  The changes to the tests include some re-organization to make it easier to accommodate expected development of Tk 9, particularly with respect to the new TclObject of type "pixels" used in Tk 9 for options whose values represent screen distances.

Please also see the discussion on PR #124112.

Note that, unlike #124112, this PR targets Python 3.14.

* Issue: gh-124111

### CODE REVIEW

File: PCbuild/tcltk.props
Comment (zooba) [inline PCbuild/tcltk.props]: Are we definitely dropping 8.6 support with this PR? Or do these need to be conditional?

File: PCbuild/tcltk.props
Comment (zooba) [inline PCbuild/tcltk.props]: This should match the version in `get_externals.bat`. It can be overridden with the `TclVersion` environment variable if you're doing a custom build.

**Review by zooba — COMMENTED** [review]
Do you know why your generated files look like they're completely changed? It is line endings being bad? (I believe our regen scripts assume you're using autocrlf on Windows.)

File: Lib/test/test_tkinter/test_misc.py
Comment (zware) [inline Lib/test/test_tkinter/test_misc.py]: Since we're already using `assertRaisesRegex` here, I think we'd be better off just adjusting the regex rather than maintaining two separate sets of messages:
```suggestion
        with self.assertRaisesRegex(TclError, r"can('| ?no)t find busy window"):
```

Similar elsewhere.

**Comment by culler** [comment]
> Do you know why your generated files look like they're completely changed? It is line endings being bad? (I believe our regen scripts assume you're using autocrlf on Windows.)

I have no idea.  I committed those by mistake, since I am used to using git commit -a.  It has something to do with Windows - those files were not changed by other
platforms.

File: Lib/test/test_tkinter/test_widgets.py
Comment (zware) [inline Lib/test/test_tkinter/test_widgets.py]: This is pretty much a styling nit, but this might be better done with a base set that is extended if tk is too new:

```suggestion
    _no_round = {'padx', 'pady'}
    _clipped = {'highlightthickness'}
    if tk_version >= (9, 0):
        _no_round.update(
            {
                'borderwidth',
                'height',
                'highlightthickness',
                'width',
            }
        )
        _clipped.update(_no_round)
```

Or as a single

```
if tk_version < (9, 0):
    # set stuff
else:
    # set stuff
```

branch.  The two ternaries back-to-back aren't especially readable to me :)

File: Lib/test/test_tkinter/widget_tests.py
Comment (zware) [inline Lib/test/test_tkinter/widget_tests.py]: Similar to above, `checkInvalidParam` uses `assertRaisesRegex` internally:

```suggestion
                errmsg=r"""image "spam" does( no|n')t exist""")
```

File: PCbuild/tcltk.props
Comment (culler) [reply PCbuild/tcltk.props]: I don't know what is the best choice for how to control whether _tkinter is built with Tk 8.6.15 or Tk 9.0.0. 

The way that I created externals/tcltk-9.0.0.0 was to build Tcl 9.0 and Tk 9.0 using nmake in the win directories of those two distributions.  The commands needed were:
`nmake -f makefile.vc`
`nmake -f makefile.vc INSTALLDIR=C\path\to\externals\tcltk-9.0.0.0\amd64 install`

I realize that to make this work in its final form tcltk-9.0.0.0 should be provided by get_externals.bat and there should be some sort of switch that controls which version gets used by build.bat.  Finalizing that design is out of the scope of this PR and beyond my expertise.  This PR is meant to be about changes to _tkinter.c and the test scripts which will be needed to make them work with both TclTk 8.6.15 and 9.0.0.

File: Lib/test/test_ttk/test_widgets.py
Comment (zware) [inline Lib/test/test_ttk/test_widgets.py]: This is unfortunate :(

Would you mind to leave a direct link to the bug in a comment here?  And do you have any expectation of when/if it may be fixed, so that we could put an upper bound on the skipped versions?  It doesn't have to be an accurate prediction; we can update the upper bound if it's missed.  I just don't want us to forget about these skips forever :)

File: PCbuild/tcltk.props
Comment (culler) [reply PCbuild/tcltk.props]: We are definitely not dropping 8.6 support with this PR.  It is meant to support both 8.6 and 9.0.  So, yes, they need to be conditional.

**Review by zware — COMMENTED** [review]
Looking good!  I have some comments that are mostly style nits, and I think we may want to hold off on the changes to `PCbuild/`.  I don't think we'll be able to move the Windows builds for <=3.13 to Tcl/Tk 9 (at least not immediately), so it will make backporting the other changes easier to leave the 3.14 `PCbuild/` updates to a separate PR.

Thanks!

File: Lib/test/test_tkinter/test_misc.py
Comment (culler) [reply Lib/test/test_tkinter/test_misc.py]: Leaving the PCbuild changes for another PR is fine with me.  I needed to test on Windows.  So I needed to build
_tkinter on Windows.  I was trying to leave some breadcrumbs so others would know what is involved in building against Tk 9 on Windows.  I guess I didn't choose the best way to do that.

Comment (zware) [reply Lib/test/test_tkinter/test_misc.py]: Those are valuable breadcrumbs that we don't want to lose :).  I do want us to update the Windows build to 9.0 as soon as we reasonably can, but we'll need to decide whether we're jumping there forever or maintaining support for 8.6 in `PCbuild/` that we ourselves won't be using.  (Personally, I think I'm fine with the direct jump in 3.14, but this is mostly going to be up to @zooba's decision.)

Comment (culler) [reply Lib/test/test_tkinter/test_misc.py]: Here is the issue with the regexes, as far as I am concerned.  Unfortunately, there have been many minor changes to the Tk error messages made recently, along the lines of can't -> can not.  That seems to be ongoing.  Maybe it will be easy to write conditional regexes which accept all of the variants.  I wasn't sure.  So I thought that having a catalog might be simpler than relying on regex cleverness.  I agree that, so far, regexes are easily able to accommodate the minor changes to error messages.  I don't know how long the error message tinkering will persist for.

A related topic, which I ignored for this PR, is that the Tcl/Tk project made a big effort to remove the "master-slave" terminology from Tcl and Tk, with considerable internal controversy.  However, tkinter is still using master and slave.  It could be using parent and child instead and be more consistent with Tcl/Tk.

File: Lib/test/test_tkinter/test_widgets.py
Comment (culler) [reply Lib/test/test_tkinter/test_widgets.py]: I wish I knew whether I could count on _no_round being a subset of _clipped.  I think the cases might be more robust at the expense of being more verbose.

File: Lib/test/test_ttk/test_widgets.py
Comment (culler) [reply Lib/test/test_ttk/test_widgets.py]: I can add a link.  But it turns out not to be a bug, exactly.  There is a change in Tk 9, which is that the combobox menu became a native macOS menu.  It looks much better and it works fine in actual human interaction with the combobox.  But the act of posting a macOS menu causes Apple to create a special event loop which does not pass events to the Tk client until the user dismisses the menu with a mouse click, which may or may not select a menu item.  While this is not an issue with using a combobox, it does create an issue with trying to use generated Tk events to open and close the combobox, as that test does.  The generated events are received and handled by Tk, but Tk currently has no way to dismiss a native macOS menu.  So it has to wait until Apple decides to dismiss the menu based on user interaction.

The choice of not having a "menu unpost" command in the macOS and Window ports of Tk was made intentionally, because those OS's provide their own (user interaction based) ways of dismissing a menu.  X11 provides an unpost command, so Tk wraps that.  In fact, macOS could support a "menu unpost" command.  But that would require a TIP since it would be a change in the Tk API.  If there were such a command then it could be bound to a Tk event and that would allow posting and dismissing combobox menus with event generate commands.

As usual, things that seem simple on the surface can be complex.

Anyway, I think I am going to close the Tk ticket as a "won't fix right now" and consider whether to write a TIP. to add a menu unpost command for Aqua.  I think it just means that that one test will have to be skipped for Tk 9.  I don't see any immediate workaround.

File: PCbuild/tcltk.props
Comment (zooba) [reply PCbuild/tcltk.props]: Overriding `TclVersion` and `tcltkDir` is the way, but changing the default `TclVersion` as you've done in the PR will break everyone else until we switch fully to relying on 9.0. So I suggest keeping it at 8.6 for now, override it yourself for testing, and when we're ready to switch to 9.0 by default then we can update it (there are more things that have to change before that's possible, so I wouldn't expect it to be in this PR).

File: Lib/test/test_tkinter/test_widgets.py
Comment (culler) [reply Lib/test/test_tkinter/test_widgets.py]: Addressed in commit 5a954ecffc.

File: Lib/test/test_ttk/test_widgets.py
Comment (culler) [reply Lib/test/test_ttk/test_widgets.py]: Addressed in commit 5a954ecffc,

File: Lib/test/test_tkinter/test_misc.py
Comment (zooba) [reply Lib/test/test_tkinter/test_misc.py]: My current preference is to keep supporting 8.6 for a while, I'm not sure how long for. But since 9.0 isn't officially released yet we certainly can't stop supporting 8.6 yet, and even after they do then we'll want to make sure it's stable and compatible enough for all people who may be building from source (which includes myself, and while we haven't thoroughly audited the 8.6 sources, arguably we should have - being forced into upgrading to 9.0 and re-auditing the sources is a cost that we'd definitely want to defer).

Comment (culler) [reply Lib/test/test_tkinter/test_misc.py]: Changes to PCbuild have been reverted.  I renamed the modified versions and added them to the repo, as examples.

File: Lib/test/test_tkinter/widget_tests.py
Comment (culler) [reply Lib/test/test_tkinter/widget_tests.py]: I removed the message catalog and replaced it with regexes, where appropriate, or cases based on tk_version.  There
weren't that many places where it was used (yet! - fingers crossed that messing with messages is over).

**Comment by culler** [comment]
I think I have now addressed all issues raised in the comments.  Is there anything else that is needed?

File: Lib/test/test_tkinter/test_widgets.py
Comment (zware) [inline Lib/test/test_tkinter/test_widgets.py]: Nit: let's not stop here :)

```suggestion
    IntegerSizeTests,
    PixelSizeTests,
)
```

File: Lib/test/test_ttk/test_widgets.py
Comment (zware) [inline Lib/test/test_ttk/test_widgets.py]: Should be switched to check against `(9, 0)` rather than `(8, 7)`?

File: Lib/test/test_tkinter/widget_tests.py
Comment (zware) [inline Lib/test/test_tkinter/widget_tests.py]: Nit: `checkInvalidParam` uses `assertRaisesRegex`, so we should be able to simplify:
```suggestion
            errmsg = 'font "" does ?n[o\']t exist'
```

**Comment by zooba** [comment]
We don't want example files in the build directory. To "save" the changes for later, they could either be a draft PR, comments in the file (not preferred), or you could add the changes with `Condition="$(TclMajorVersion) == '9'"` checks so that they automatically apply when building against 9.0 (preferred).[^1]

[^1]: Number comparisons are a bit wonky in MSBuild, so usually just best to pick the most sensible default. In this case, it's probably actually to check `$(TclMajorVersion) == '8'` for the old behaviour, so that when 10.0 comes out it'll get the newer behaviours.

**Comment by culler** [comment]
> We don't want example files in the build directory. To "save" the changes for later, they could either be a draft PR, comments in the file (not preferred), or you could add the changes with `Condition="$(TclMajorVersion) == '9'"` checks so that they automatically apply when building against 9.0 (preferred).[1](#user-content-fn-1-9d9cf4f3af8602fa729bad031ef1ba2c)
> ## Footnotes
> 
>     1. Number comparisons are a bit wonky in MSBuild, so usually just best to pick the most sensible default. In this case, it's probably actually to check `$(TclMajorVersion) == '8'` for the old behaviour, so that when 10.0 comes out it'll get the newer behaviours. [↩](#user-content-fnref-1-9d9cf4f3af8602fa729bad031ef1ba2c)

Sure.  I will attempt to modify _tkinter.vcxproj to allow building against either 8.6.14 or 9.0.0.  (Note that 8.6.15 has been released, but the binaries are not available to get_externals.bat yet.)

File: Lib/test/test_ttk/test_widgets.py
Comment (zware) [inline Lib/test/test_ttk/test_widgets.py:1195]: Trying this locally on Linux with Tcl/Tk 9.0b3, I get a failure where 'focus' is gotten instead of 'padding'.  Maybe we should just adjust the test like the following?

```suggestion
        self.assertIn(self.nb.identify(5, 5), ['', 'focus', 'padding'])
```

**Comment by culler** [comment]
I have removed the "examples" and modified the build scripts to use the TclVersion environment variable to decide which Tk version gets used for the build.  (The version must also be installed in ../externals/tcltk-W.X.Y.Z, which can either be done locally or by the get_externals script.  I tested on my machine and was able to build both for Tk 8.6.14 (using the binaries downloaded by get_externals) and for Tk 9.0.0.0 (installed by me, by setting the INSTALLDIR variable in a standaard build of Tcl and Tk with nmake and makefile.vc).

**Review by zware — COMMENTED** [review]
Some of these changes cause failures for me on Pop_OS! 22.04 (Ubuntu 22.04 derivative) with Tcl/Tk 9.0.0b3.  It's possible that I'm getting my configuration wrong, but switching to Tcl/Tk 9.0.0b3 without this PR causes one failure, then adding these changes introduces a bunch of other failures, mostly around rounding as far as I can tell.

File: Lib/test/test_ttk/test_widgets.py
Comment (culler) [reply Lib/test/test_ttk/test_widgets.py]: I left quite a few tests against (8, 7) in place.  There are many more than just this.  They are not wrong.  They are just useless, assuming that no one ever tries to support using 8.7 with tkinter.  Our understanding is that 8.7 will never be released, so that  assumption is presumably valid.  I guess I could go through all files and replace (8,7) by (9,0) if that seems worth doing.

File: Lib/test/test_tkinter/test_widgets.py
Comment (culler) [reply Lib/test/test_tkinter/test_widgets.py]: OK

File: Lib/test/test_tkinter/widget_tests.py
Comment (culler) [reply Lib/test/test_tkinter/widget_tests.py]: OK

File: Lib/test/test_ttk/test_widgets.py
Comment (culler) [reply Lib/test/test_ttk/test_widgets.py:1195]: I think this is due to a recent change in the size of the focus ring.

**Comment by culler** [comment]
> Some of these changes cause failures for me on Pop_OS! 22.04 (Ubuntu 22.04 derivative) with Tcl/Tk 9.0.0b3. It's possible that I'm getting my configuration wrong, but switching to Tcl/Tk 9.0.0b3 without this PR causes one failure, then adding  these changes introduces a bunch of other failures, mostly around rounding as far as I can tell.

I am getting no failures on Ubuntu 22.04.  I think that you are probably using the released 9.0.b3 while I am using the tips of main branches of the fossil repositories for Tcl and Tk.  (I did that because they presumably are closer to the actual 9.0.0 releases.)  There have been recent changes to how some of the widgets handle rounding.

I am using commits 8627aad9bd for Tcl and d2f53df7a8 for Tk. (Both from 9/23)

Note: part of what I am trying to do here is to make it easier to adjust the tests as these sorts of changes happen.  I think the plan is that eventually all of these options will be rounded.  Also more options will be changed to accept pixel values (e.g. 10p or 23m).

**Comment by zware** [comment]
Ahh, that makes sense.  It also makes me think we probably want to sit on this until 9.0 is actually released or has at least had a release candidate, though :-/

**Comment by culler** [comment]
> Ahh, that makes sense. It also makes me think we probably want to sit on this until 9.0 is actually released or has at least had a release candidate, though :-/

Sure.  A release candidate is imminent.  The [core-9-0-0-rc](https://core.tcl-lang.org/tk/timeline?r=core-9-0-0-rc) branch is being finalized right now.

**Comment by culler** [comment]
I switched to the tip of the core-9-0-0-rc branches of Tcl and Tk and reran the tkinter and ttk tests on Ubuntu 22.04,  I still am seeing no errors.

File: Lib/test/test_tkinter/test_misc.py
Comment (zware) [inline Lib/test/test_tkinter/test_misc.py]: Nit: dangling unused import:

```suggestion
from test.test_tkinter.support import AbstractTkTest, AbstractDefaultRootTest, requires_tk
```

**Comment by culler** [comment]
I think this PR is finished.  The PR checks test against Tcl/Tk 8.6.14, and they are showing no failures.  I have tested against Tcl/Tk 8.6.15 and Tcl/Tk 9.0.0 (rc2)  on my own machines, running Ubuntu 22.04, Windows 11, and macOS 14.  There are no failures in those 6 cases.  The issue reported in #124378 with 8.6.15 has been resolved here.

The release of Tcl/Tk 9.0.0 is expected within a week, and surely will happen by the end of next week.

File: Lib/test/test_tkinter/support.py
Comment (zware) [inline Lib/test/test_tkinter/support.py]: Nit: this doesn't seem to be used anymore.

**Review by zware — APPROVED** [review]
This is great, thank you!  I'm satisfied with this change overall, but I think we should get a once-over from @serhiy-storchaka for the tkinter changes and a final confirmation from @zooba that the `PCbuild/` changes match his expectations.

With your permission, I'd like to extract your ttk Notebook test fixes to a separate PR so that we can go ahead with it and get it backported.  (If you would prefer to make that split yourself, that works for me.)

I expect that we will eventually be backporting this whole PR even if we don't start shipping 9.0 with our <3.14 binaries because at some point our non-Windows CI systems will inevitably update to 9.0, but I'm not sure how quickly that will happen and want to be sure we get 8.6.15 supported as soon as we can.

File: Lib/test/test_tkinter/test_misc.py
Comment (culler) [reply Lib/test/test_tkinter/test_misc.py]: I pushed one more commit containing these changes and the ones above.

**Comment by culler** [comment]
According to [this email](https://sourceforge.net/p/tcl/mailman/message/58821470/) the release of Tcl/Tk 9.0.0.0 will happen tomorrow.

File: PCbuild/_tkinter.vcxproj
Comment (zooba) [inline PCbuild/_tkinter.vcxproj]: Do we have to redistribute this file with Tcl/Tk 9.0? If so, there'll be updates under `Tools/msi` as well (look for references to tclZlibDllName).

Also, we should be using the name property here rather than the filename directly.

File: PCbuild/tcltk.props
Comment (zooba) [inline PCbuild/tcltk.props]: These are names, not paths, so don't include the path.

```suggestion
    <tommathDLLName Condition="'$(TclMajorVersion)' == '9'">libtommath.dll</tommathDLLName>
    <tommathLibName Condition="'$(TclMajorVersion)' == '9'">tommath.lib</tommathLibName>
```

Add the path in the tcltkLib setting, or potentially just add `$(tcltkDir)Lib` to `_tkinter.vcxproj` as `AdditionalLibraryDirectories` and then the tcltkLib property can just contain filenames.

File: PCbuild/_tkinter.vcxproj
Comment (zooba) [reply PCbuild/_tkinter.vcxproj]: Once we're using the name property, checking whether the name is empty rather than checking the version is going to be more robust.

```suggestion
    <_TclTkDLL Include="$(tcltkdir)\bin\$(tommathDLLName)" Condition="$(tommathDLLName) != ''"/>
```

File: PCbuild/tcltk.props
Comment (zooba) [inline PCbuild/tcltk.props:20]: This is not used anywhere, just remove it.

**Review by zooba — COMMENTED** [review]
Along with the MSI changes (and possibly under PC/layout as well, I haven't checked), we need to make sure that we're okay with redistributing libtommath.dll and also that there isn't a static linking option for it. zlib unfortunately causes us trouble, but static linking here is impossible (because Tcl/Tk doesn't allow for it in their build scripts - I suspect the same will apply to tommath).

I'm okay with merging without the MSI changes, but we then cannot update any releases until it's done. As annoying as it is for backports, I'd require a comment in get_externals.bat to this effect.

File: PCbuild/_tkinter.vcxproj
Comment (culler) [reply PCbuild/_tkinter.vcxproj]: Now using names, not paths and checking whether the name is empty in the Condition.

File: PCbuild/tcltk.props
Comment (culler) [reply PCbuild/tcltk.props:20]: Actually, the TCL_WITH_EXTERNAL_TOMMATH constant **is** used when building against TclTk 9.0; although it is not used by 8.6.  With 9.0, if it is not defined then all of the symbols defined in libtommath.dll are reported by the linker as being undefined.  I believe that this is related to how the Tcl stubs library works;  it is not a straightforward linker error since libtommath.dll is being passed to the linker.

One might hope that the existence of a constant named TCL_WITH_EXTERNAL_TOMMATH indicates that one could choose not to have an external libtommath.dlll and instead arrange for libtommath to be statically linked into libtcl.dll, as is the case with Tcl 8.6.  However, I was not able to find a combination of build options which accomplished that.  I will try to find out if it is possible.

File: PCbuild/tcltk.props
Comment (culler) [reply PCbuild/tcltk.props]: I ended up adding the path in tcltkLib.  It did not seem to work to add it to AdditionalLibraryDirectories, although I don't know why not.

File: PCbuild/tcltk.props
Comment (culler) [reply PCbuild/tcltk.props:20]: @zooba: As far as I can tell. the following command uses MSVC to build a static library tcl90s.lib which has both zlib and (part of) libtommath built-in:
 `nmake -f makefile.vc TCL_WITH_INTERNAL_ZLIB=1 TCL_WITH_EXTERNAL_TOMMATH=0 STATIC_BUILD=1`

According to dumpbin, the tclsh90s.exe executable built by that command does not have libtommath.dll or zlib.dll as dependencies.  So presumably the same would be true of the _tkinter extension if it were linked against tcl90s.lib. 

If you are willing to use gcc in msys64 then `configure --disable-static` seems to do the equivalent thing.

Reference: [https://core.tcl-lang.org/tips/doc/trunk/tip/538.md](https://core.tcl-lang.org/tips/doc/trunk/tip/538.md)

**Comment by oehhar** [comment]
Big applause to Marc for this big bunch of work. I am sure, the TkInter community will enjoy tk 9.0, as it has a better look and scaling ability of the whole gui.
Harald with greeting from the Tcl/Tk team !

**Comment by vstinner** [comment]
I propose to backport this change to Python 3.13, see: https://github.com/python/cpython/pull/127364.

See the discussion on the issue for the backport rationale.

**Pull request merged**
